### PR TITLE
Deseq2

### PR DESCRIPTION
**Pull request by rocherroche**

### CODE REVIEW

File: pegasus/tools/pseudobulk.py [hunk 1]
```diff
@@ -144,3 +148,83 @@ def pseudobulk(
...
+            dds = de_seq.DESeq(dds, minReplicatesForReplace=math.inf)
+            res= de_seq.results(dds, contrast=robjects.StrVector(contrast), cooksCutoff=False)
+
+        with localconverter(ro.default_converter + pandas2ri.converter):
+          res_df = ro.conversion.rpy2py(to_dataframe(res))
+          res_df.fillna({'log2FoldChange': 0.0, 'lfcSE': 0.0, 'stat': 0.0, 'pvalue': 1.0, 'padj': 1.0}, inplace=True)
+
+        de_res_key = de_key if mat_key.find('.') < 0 else f"{mat_key.partition('.')[0]}.{de_key}"
+        pseudobulk.varm[de_res_key] = res_df.to_records(index=False)
+        pandas2ri.deactivate()
```
Comment (yihming) [inline pegasus/tools/pseudobulk.py]: Should this line be within the for loop or outside?

**Review by yihming — CHANGES_REQUESTED** [review]

File: pegasus/tools/clustering.py [hunk 2]
```diff
@@ -177,8 +177,8 @@ def louvain(
     try:
         import louvain as louvain_module
     except ImportError:
-        print("Need louvain! Try 'pip install louvain-github'.")
-
+        logger.error("Need louvain! Try 'pip install louvain-github'.")
```
Comment (bli25) [inline pegasus/tools/clustering.py:181]: Please use import sys; sys.exit(-1)

File: pegasus/tools/clustering.py [hunk 3]
```diff
@@ -251,7 +251,8 @@ def leiden(
     try:
         import leidenalg
     except ImportError:
-        print("Need leidenalg! Try 'pip install leidenalg'.")
+        logger.error("Need leidenalg! Try 'pip install leidenalg'.")
```
Comment (bli25) [inline pegasus/tools/clustering.py:256]: Please use import sys; sys.exit(-1)

File: pegasus/tools/clustering.py [hunk 4]
```diff
@@ -380,7 +381,8 @@ def spectral_louvain(
     try:
         import louvain as louvain_module
     except ImportError:
-        print("Need louvain! Try 'pip install louvain-github'.")
+        logger.error("Need louvain! Try 'pip install louvain-github'.")
```
Comment (bli25) [inline pegasus/tools/clustering.py:387]: Please use import sys; sys.exit(-1)

File: pegasus/tools/clustering.py [hunk 5]
```diff
@@ -479,7 +481,8 @@ def spectral_leiden(
     try:
         import leidenalg
     except ImportError:
-        print("Need leidenalg! Try 'pip install leidenalg'.")
+        logger.error("Need leidenalg! Try 'pip install leidenalg'.")
```
Comment (bli25) [inline pegasus/tools/clustering.py:488]: Please use import sys; sys.exit(-1)

File: pegasus/tools/gradient_boosting.py [hunk 6]
```diff
@@ -92,7 +92,8 @@ def find_markers(
     try:
         from lightgbm import LGBMClassifier
     except ImportError:
-        print("Need lightgbm! Try 'pip install lightgbm'.")
+        logger.error("Need lightgbm! Try 'pip install lightgbm'.")
```
Comment (bli25) [inline pegasus/tools/gradient_boosting.py:96]: Please use import sys; sys.exit(-1)

File: pegasus/tools/graph_operations.py [hunk 7]
```diff
@@ -1,10 +1,13 @@
...
 import numpy as np
 from scipy.sparse import issparse, csr_matrix
+
+import logging
+logger = logging.getLogger(__name__)
 try:
     import igraph
 except ImportError:
-    print("Need python-igraph!")
+    logger.error("Need python-igraph!  Try 'pip install python-igraph'.")
```
Comment (bli25) [inline pegasus/tools/graph_operations.py:10]: Please use import sys; sys.exit(-1)

File: pegasus/tools/pseudotime.py [hunk 8]
```diff
@@ -1,10 +1,13 @@
...
 import numpy as np
 from sklearn.metrics.pairwise import euclidean_distances
+
+import logging
+logger = logging.getLogger(__name__)
 try:
     import igraph
 except ImportError as error:
-    print("Need python-igraph!")
+    logger.error("Need python-igraph!  Try 'pip install python-igraph'.")
```
Comment (bli25) [inline pegasus/tools/pseudotime.py:10]: Please use import sys; sys.exit(-1)

File: pegasus/tools/visualization.py [hunk 9]
```diff
@@ -56,8 +56,8 @@ def calc_tsne(
     try:
         from fitsne import FItSNE
     except ModuleNotFoundError:
-        print("Exception: FItSNE package is not installed, please install fitsne: pip install fitsne")
-        exit()
+        logger.error("Need FItSNE!  Try 'pip install fitsne'.")
```
Comment (bli25) [inline pegasus/tools/visualization.py:59]: Please use import sys; sys.exit(-1)

File: pegasus/tools/visualization.py [hunk 10]
```diff
@@ -142,8 +142,8 @@ def calc_force_directed_layout(
     try:
         import forceatlas2 as fa2
     except ModuleNotFoundError:
-        print("Exception: Please install module forceatlas2-python: pip install forceatlas2-python")
-        exit()
+        logger.error("Need forceatlas2-python!  Try 'pip install forceatlas2-python'.")
```
Comment (bli25) [inline pegasus/tools/visualization.py:146]: Please use import sys; sys.exit(-1)

**Review by bli25 — CHANGES_REQUESTED** [review]

**Review by bli25 — APPROVED** [review]

**Pull request merged**